lockfile/src/parse_depfile.rs

```diff
@@ -0,0 +1,249 @@
+    #[error("Parsing {0:?} requires lockfile generation, but it was disabled")]
+    ManifestWithoutGeneration(String),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+/// Lockfile with all its dependencies.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ParsedLockfile {
+    pub path: String,
+    pub packages: Vec<PackageDescriptor>,
+    pub format: LockfileFormat,
+}
+
+impl ParsedLockfile {
+    pub fn new(
+        path: impl Into<String>,
+        format: LockfileFormat,
+        packages: Vec<PackageDescriptor>,
+    ) -> Self {
+        Self { path: path.into(), packages, format }
+    }
+
+    /// Convert packages to API's expected format.
+    pub fn api_packages(&self) -> Vec<PackageDescriptorAndLockfile> {
+        self.packages
+            .iter()
+            .map(|package_descriptor| PackageDescriptorAndLockfile {
+                package_descriptor: package_descriptor.clone(),
+                lockfile: Some(self.path.clone()),
+            })
+            .collect()
+    }
+}
+
+/// Parse a dependency file.
+///
+/// The `path` argument is used for format identification and display purposes.
+///
+/// The `_generation_path` must point to the manifest on the filesystem if
+/// lockfile generation should be performed. Use `None` to disable lockfile
+/// generation.
+///
+/// # Features
+///
+/// Parsing manifests requires the `generator` feature.
+pub fn parse_depfile(
+    contents: &str,
+    path: impl Into<String>,
+    format: Option<LockfileFormat>,
+    _generation_path: Option<PathBuf>,
+) -> Result<ParsedLockfile, ParseError> {
+    // Try to determine the dependency file format.
+    let path = path.into();
+    let format = format.or_else(|| crate::get_depfile_path_format(&path));
+
+    // Attempt to parse with all known parsers as fallback.
+    let format = match format {
+        Some(format) => format,
+        None => return Ok(try_get_packages(path, contents)?),
+    };
+
+    // Parse with the identified parser.
+    let parser = format.parser();
+
+    // Check if file could be a lockfile/manifest based on file name.
+    let pseudopath = PathBuf::from(&path);
+    let maybe_lockfile = parser.is_path_lockfile(&pseudopath);
+    let maybe_manifest = parser.is_path_manifest(&pseudopath);
+
+    // Attempt to parse the identified lockfile.
+    let mut lockfile_error = None;
+    if maybe_lockfile || !maybe_manifest {
+        // Parse lockfile content.
+        let packages = parse_lockfile_content(contents, parser);
+
+        match packages {
+            Ok(packages) => return Ok(ParsedLockfile::new(path, format, packages)),
+            // Store error on failure.
+            Err(err) => lockfile_error = Some(err),
+        }
+    }
+
+    // Generate lockfile if path might be a manifest and feature and option are
+    // enabled.
+    #[cfg(feature = "generator")]
+    if let Some(generation_path) = _generation_path {
```

Comment:
We still need to check `maybe_manifest`... Otherwise, we'll attempt lockfile generation on all sorts of random files.

```
> echo "BAD" > Cargo.lock; ./target/debug/phylum parse Cargo.lock
❗ Error: could not parse dependency file "Cargo.lock"

Caused by:
    0: Dependency file parsing failed
    1: subprocess failed:

[ ... SNIP (config errors, etc. addressed in separate comments) ...]

       Generating lockfile for manifest "Cargo.lock" using Cargo…
       ❗ Error: Lockfile generation failed! For details, see: https://docs.phylum.io/docs/lockfile_generation

       Caused by:
           package manager quit unexpectedly (code: Some(101)):

           error: the manifest-path must be a path to a Cargo.toml file
```

```suggestion
    if let Some(generation_path) = _generation_path.filter(|_| maybe_manifest) {
```